### PR TITLE
[runtime/trace] create stackwalker

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

**Comment by isuruf** [comment]
Added you to the appveyor team

**Comment by ghost** [comment]
@isuruf I think cmake is passing the fortran files `-DLINUX` and such. I tried to kill that but failed. Do you have any ideas there?

**Comment by ghost** [comment]
Also, heres, `_fortio_binary_mode`:

```
int
__io_binary_mode(void *fp)
{
#if defined(WINNT)
#include <fcntl.h>

#if defined(WIN64) || defined(WIN32)
#define O_BINARY _O_BINARY
#endif

 int mode;

 mode = setmode(fileno((FILE *)fp), O_BINARY);
 if (mode == -1) {
   /* The mode argument is clearly legal, so this should not
    * happen.  But, in a console app, setmode will fail on
    * the fd representing stdout.
    */
   return 0;
 }
 (void)setmode(fileno((FILE *)fp), mode);
 return (mode & O_BINARY);
#else
 return 1;
#endif
}
```

**Comment by isuruf** [comment]
It's done here, https://github.com/flang-compiler/clang/blob/flang_release_40/lib/Driver/Tools.cpp#L4702

**Comment by ghost** [comment]
That's not what I'm talking about. This is nmake output:

```
C:\Miniconda36-x64\Library\bin\flang.exe -DGTEST_HAS_RTTI=0 -DHOST_WIN -DLINUX -DMAXCPUS=256 -DMAXCPUSL=8 -DMAXCPUSR=8 -DNATIVE_FPCVT -DPGF90 -DPGFLANG -DPGI_LITTLE_ENDIAN -DTARGET_LINUX -DTARGET_WIN -DTARGET_WIN_X86 -DTARGET_WIN_X8664 -DTARGET_X8664 -DUNICODE -DWIN32 -DWIN64 -DWINNT -D_CRT_NONSTDC_NO_DEPRECATE -D_CRT_NONSTDC_NO_WARNINGS -D_CRT_SECURE_NO_DEPRECATE -D_CRT_SECURE_NO_WARNINGS -D_DEBUG_POINTER_IMPL="" -D_GNU_SOURCE -D_HAS_EXCEPTIONS=0 -D_SCL_SECURE_NO_DEPRECATE -D_SCL_SECURE_NO_WARNINGS -D_UNICODE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Dflang_shared_EXPORTS -DHAVE_LONG_LONG_INT -DINT32PTR64 -DTARGET_WIN_X8664 -DTM_I8 -IC:\projects\flang\include -IC:\projects\flang\build\include -IC:\Miniconda36-x64\Library\include -IC:\projects\flang\runtime\include -IC:\projects\flang\runtime\flang -IC:\projects\flang\build\runtime\flang -B C:/projects/flang/build/./bin    -Mreentrant -c C:\projects\flang\runtime\flang\vmmul_real8.F95 -o CMakeFiles\flang_shared.dir\vmmul_real8.F9
```

File: runtime/flangrti/trace/StackWalker.cpp [hunk 1]
```diff
@@ -0,0 +1,1365 @@
...
+
+    char buf[StackWalker::STACKWALK_MAX_NAMELEN] = {0};
+    if (this->pSGSP != NULL)
+    {
+      if (this->pSGSP(m_hProcess, buf, StackWalker::STACKWALK_MAX_NAMELEN) == FALSE)
+        this->m_parent->OnDbgHelpErr("SymGetSearchPath", GetLastError(), 0);
+    }
+    char szUserName[1024] = {0};
+    DWORD dwSize = 1024;
+    GetUserNameA(szUserName, &dwSize);
```
Comment (isuruf) [inline runtime/flangrti/trace/StackWalker.cpp:406]: This requires advapi32.lib, see https://msdn.microsoft.com/en-us/library/ms724432(VS.85).aspx#Mtps_DropDownFilterText

**Comment by ghost** [comment]
@isuruf this approach  is completely wrong, should be 

```
void printStack( void )
{
     unsigned int   i;
     void         * stack[ 100 ];
     unsigned short frames;
     SYMBOL_INFO  * symbol;
     HANDLE         process;

     process = GetCurrentProcess();

     SymInitialize( process, NULL, TRUE );

     frames               = CaptureStackBackTrace( 0, 100, stack, NULL );
     symbol               = ( SYMBOL_INFO * )calloc( sizeof( SYMBOL_INFO ) + 256 * sizeof( char ), 1 );
     symbol->MaxNameLen   = 255;
     symbol->SizeOfStruct = sizeof( SYMBOL_INFO );

     for( i = 0; i < frames; i++ )
     {
         SymFromAddr( process, ( DWORD64 )( stack[ i ] ), 0, symbol );

         printf( "%i: %s - 0x%0X\n", frames - i - 1, symbol->Name, symbol->Address );
     }

     free( symbol );
}
```

**Pull request merged**